src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1041,102 +1041,184 @@ private bool IsValidFixedVariableInitializer(TypeSymbol declType, SourceLocalSym
+                        }
+
+                        if (extensisbleFixedEnabled)
+                        {
+                            // not a specially known type, check for getPinnableMethod
+                            if (getPinnableMethod != null)
                             {
-                                Error(diagnostics, ErrorCode.ERR_FixedNotNeeded, initializerSyntax);
-                                return true;
+                                elementType = getPinnableMethod.ReturnType;
+                                break;
                             }
-                            else
+                            else if (additionalDiagnostics != null)
                             {
-                                elementType = ((PointerTypeSymbol)fa.Type).PointedAtType;
+                                diagnostics.AddRange(additionalDiagnostics);
                             }
-                            break;
-                        case BoundKind.Conversion:
-                            // The following assertion would not be correct because there might be an implicit conversion after (above) the explicit one.
-                            //Debug.Assert(((BoundConversion)initializerOpt).ExplicitCastInCode, "The assignment conversion hasn't been applied yet, so this must be from source.");
-
-                            // NOTE: Dev10 specifically doesn't report this error for the array or string cases.
-                            Error(diagnostics, ErrorCode.ERR_BadCastInFixed, initializerSyntax);
-                            return true;
-                        default:
-                            // CONSIDER: this is a very confusing error message, but it's what Dev10 reports.
-                            Error(diagnostics, ErrorCode.ERR_FixedNotNeeded, initializerSyntax);
-                            return true;
+                        }
+                        else if (getPinnableMethod != null)
+                        {
+                            CheckFeatureAvailability(initializerOpt.Syntax, MessageID.IDS_FeatureExtensibleFixedStatement, diagnostics);
+                        }
+
+                        Error(diagnostics, ErrorCode.ERR_ExprCannotBeFixed, initializerSyntax);
                     }
+                    finally
+                    {
+                        additionalDiagnostics.Free();
+                    }
+
+                    return false;
+            }
+
+            if (elementType.IsManagedType)
+            {
+                Error(diagnostics, ErrorCode.ERR_ManagedAddr, initializerSyntax, elementType);
+                hasErrors = true;
+            }
+
+            initializerOpt = GetFixedLocalCollectionInitializer(initializerOpt, elementType, declType, getPinnableMethod, hasErrors, diagnostics);
+            return true;
+        }
+
+        private MethodSymbol GetDangerousGetPinnableReferenceMethod(BoundExpression initializer, DiagnosticBag additionalDiagnostics)
```

Comment:
Nit: no need for "Dangerous" here

Reply:
I think we are not keeping "Dangerous" in the name. Once I implement the name change this will be renamed too

---
In reply to: [167647561](https://github.com/dotnet/roslyn/pull/24469#discussion_r167647561) [](ancestors = 167647561)